### PR TITLE
Fixes #11413 - Conscrypt does not support server-side SNI.

### LINKED ISSUE #11413: Conscrypt does not support server-side SNI
**Jetty version(s)**
12.0.7

**Jetty Environment**
core

**Java version/vendor** `(use: java -version)`
openjdk version "21.0.1" 2023-10-17 LTS
OpenJDK Runtime Environment Zulu21.30+15-CA (build 21.0.1+12-LTS)
OpenJDK 64-Bit Server VM Zulu21.30+15-CA (build 21.0.1+12-LTS, mixed mode, sharing)

**OS type/version**
Microsoft Windows [Version 10.0.19045.4046]

**Description**
https://eclipse.dev/jetty/documentation/jetty-12/operations-guide/index.html#og-protocols-ssl-sni states that when `sniRequired=false` the SNI checks should always pass.

If you take a look at the attached DEBUG log, you'll notice that `sniRequired=false` (the default setting. I didn't alter it) but the server still responds with `org.eclipse.jetty.http.BadMessageException: 400: Invalid SNI`.

[debug.log](https://github.com/jetty/jetty.project/files/14303424/debug.log)

I suspect I know what is going wrong in my particular case: the browser is hitting the server by IP address, and that IP address happens to be my own IP address. I'm not sure if that ends up getting rerouted to localhost but it could be the case.

Per https://stackoverflow.com/a/69945374/14731:

> Know that SNI itself has restrictions:
> 1. localhost is not allowed
> 2. IP Address Literals are not allowed (eg: 192.168.1.215 or fe80::3831:400c:dc07:7c40)

If the failure I am seeing is expected behavior (working as designed) then I'd suggest making the following changes:

1. Update the documentation to mention these SNI restrictions.
2. Add more information to the exception message, such as `Invalid SNI. localhost is not allowed` or `Invalid SNI IP address literals are not allowed`, etc.

Out of curiosity, is there any way to bypass these restrictions? Or do I have to hit the server by name instead of by IP?

### PR DESCRIPTION
**Pull request by sbordet**
Modified `SecureRequestCustomizer` to use `ExtendedSSLSession` APIs instead of relying on SSLSession attributes, that do not work in Conscrypt.

### CODE REVIEW

File: jetty-core/jetty-alpn/jetty-alpn-conscrypt-server/src/test/java/org/eclipse/jetty/alpn/conscrypt/server/ConscryptHTTP2ServerTest.java [hunk 1]
```diff
@@ -140,4 +139,12 @@ public void testSimpleRequest() throws Exception
             assertEquals(200, contentResponse.getStatus());
         }
     }
+
+    @Test
+    public void testSNIRequired() throws Exception
+    {
+        // The KeyStore contains 1 certificate with two DNS names.
+        httpsConfig.getCustomizer(SecureRequestCustomizer.class).setSniRequired(true);
```
Comment (lorban) [inline jetty-core/jetty-alpn/jetty-alpn-conscrypt-server/src/test/java/org/eclipse/jetty/alpn/conscrypt/server/ConscryptHTTP2ServerTest.java:147]: This modifies the `httpsConfig` instance for all tests of this class ran after this one.

A fresh `httpsConfig` should be created for each test.

**Review by lorban — CHANGES_REQUESTED** [review]
LGTM besides a small problem with the test.

File: jetty-core/jetty-alpn/jetty-alpn-conscrypt-server/src/test/java/org/eclipse/jetty/alpn/conscrypt/server/ConscryptHTTP2ServerTest.java (hunk 1 above)
Comment (sbordet) [reply jetty-core/jetty-alpn/jetty-alpn-conscrypt-server/src/test/java/org/eclipse/jetty/alpn/conscrypt/server/ConscryptHTTP2ServerTest.java:147]: No, I just checked, and JUnit allocates a new instance for every test method. This is evident also from the logs that print different hash codes for each `Server` instance:
```
...
2024-11-19 18:57:46.602:INFO :oejs.Server:main: Started oejs.Server@13bc8645{STARTING}[12.0.16-SNAPSHOT,sto=0] @824ms
...
2024-11-19 18:57:46.815:INFO :oejs.Server:main: Started oejs.Server@34f22f9d{STARTING}[12.0.16-SNAPSHOT,sto=0] @1037ms
```

**Review by lorban — DISMISSED** [review]

File: jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/SecureRequestCustomizer.java [hunk 2]
```diff
@@ -230,6 +232,31 @@ protected void checkSni(Request request, SSLSession session)
...
+        {
+            sniHost = extended.getRequestedServerNames().stream()
+                .filter(SNIHostName.class::isInstance)
+                .map(SNIHostName.class::cast)
+                .map(SNIHostName::getAsciiName)
+                .findFirst()
+                .orElse(null);
+            if (sniHost != null)
+                return sniHost;
+        }
```
Comment (gregw) [inline jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/SecureRequestCustomizer.java]: I don't think streams add any clarity here.  The code below is clearer and more efficient:

```suggestion
        if (session instanceof ExtendedSSLSession extended)
        {
            for (SNIServerName serverName : extended.getRequestedServerNames())
            {
                if (serverName instanceof SNIHostName hostName)
                    return hostName.getAsciiName();
            }
        }
```

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/ssl/SniX509ExtendedKeyManager.java [hunk 3]
```diff
@@ -115,18 +113,14 @@ protected String chooseServerAlias(String keyType, Principal[] issuers, Collecti
...
-                    .map(SNIHostName.class::cast)
-                    .map(SNIHostName::getAsciiName)
-                    .orElse(null);
-            }
+            host = extended.getRequestedServerNames().stream()
+                .filter(SNIHostName.class::isInstance)
+                .map(SNIHostName.class::cast)
+                .map(SNIHostName::getAsciiName)
+                .findFirst()
+                .orElse(null);
```
Comment (gregw) [inline jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/ssl/SniX509ExtendedKeyManager.java]: Ditto don't use streams here

**Review by gregw — CHANGES_REQUESTED** [review]

**Review by gregw — APPROVED** [review]

**Pull request merged**